pandas/core/categorical.py

```diff
@@ -1388,15 +1388,23 @@ def unique(self):
 
         Unused categories are NOT returned.
 
+        For ordered categoricals, the unique values are returned in order.
+        For unordered categoricals, the unique values are returned in order of
+        appearance.
+
         Returns
         -------
         unique values : array
         """
-        unique_codes = np.unique(self.codes)
-        # for compatibility with normal unique, which has nan last
-        if unique_codes[0] == -1:
-            unique_codes[0:-1] = unique_codes[1:]
-            unique_codes[-1] = -1
+        from pandas.core.nanops import unique1d
+        # unlike np.unique, unique1d does not sort
+        unique_codes = unique1d(self.codes)
+        if self.ordered:
```

Comment:
The more I think about this (and read the bugreports which have  problems with this), the more I disagree with this change: if unique returns values in the order in which they appear in the cols, then this "contract" should be used. "order" only says that values in the categorical can be ordered, not that unique switches the order. If one wants to have the categories in order, then take the `categories` directly (e.g. https://github.com/mwaskom/seaborn/issues/361)

The only think missing is the "get me all unique values in the same order as specified by the categories", and one could give that a new keyword or a new method